extensions/attributegen/plugin.h

```diff
@@ -0,0 +1,205 @@
+#include "extensions/attributegen/config.pb.h"
+#include "google/protobuf/util/json_util.h"
+
+// WASM_PROLOG
+#ifndef NULL_PLUGIN
+
+#include "proxy_wasm_intrinsics.h"
+// Do not reorder.
+#include "extensions/common/proxy_expr.h"
+
+#else  // NULL_PLUGIN
+
+#include "extensions/common/wasm/null/null_plugin.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+namespace Null {
+namespace Plugin {
+
+using WasmResult = Envoy::Extensions::Common::Wasm::WasmResult;
+using NullPluginRegistry =
+    ::Envoy::Extensions::Common::Wasm::Null::NullPluginRegistry;
+using Envoy::Extensions::Common::Wasm::Null::Plugin::FilterStatus;
+
+#include "api/wasm/cpp/contrib/proxy_expr.h"
+
+#endif  // NULL_PLUGIN
+
+// END WASM_PROLOG
+
+namespace AttributeGen {
+
+using StringView = absl::string_view;
+
+using google::protobuf::util::JsonParseOptions;
+using google::protobuf::util::Status;
+
+class Match {
+ public:
+  explicit Match(std::string condition, uint32_t condition_token,
+                 std::string value)
+      : condition_(condition),
+        condition_token_(condition_token),
+        value_(value){};
+
+  // Returns the result of evaluation or nothing in case of an error.
+  Optional<bool> evaluate() const {
+    if (condition_.empty()) {
+      LOG_DEBUG(absl::StrCat("Matched empty condition: value=", value_));
+      return true;
+    }
+
+    bool matched;
+    if (!evaluateExpression(condition_token_, &matched)) {
+      LOG_WARN(absl::StrCat("Failed to evaluate expression: ", condition_));
+      return {};
+    }
+    if (matched) {
+      LOG_DEBUG(absl::StrCat("Matched: ", condition_));
+    }
+    return matched;
+  };
+
+  std::string value() const { return value_; };
+
+  void cleanup() {
+    if (!condition_.empty()) {
+      exprDelete(condition_token_);
+    }
+  }
+
+ protected:
+  // Only used for debug messages after construction.
+  std::string condition_;
+  // Expression token associated with the condition.
+  uint32_t condition_token_;
+  std::string value_;
+};
+
+enum EvalPhase { on_log, on_request };
+
+class AttributeGenerator {
+ public:
+  explicit AttributeGenerator(EvalPhase phase, std::string output_attribute)
+      : phase_(phase),
+        output_attribute_(output_attribute),
+        error_attribute_(absl::StrCat(output_attribute, "_error")) {}
+
+  // If evaluation is successful returns true and sets result.
+  Optional<bool> evaluate(std::string* val) const {
+    for (const auto& match : matches_) {
+      auto eval_status = match.evaluate();
+      if (!eval_status) {
+        return {};
+      }
+      if (eval_status.value()) {
+        *val = match.value();
+        return true;
+      }
+    }
+    return false;
+  }
+
+  EvalPhase phase() const { return phase_; }
+  std::string output_attribute() const { return output_attribute_; }
+  std::string error_attribute() const { return error_attribute_; }
+  void add_match(Match match) { matches_.push_back(match); }
```

Comment:
Why does this need to be mutable? It's created once and const after that.

Reply:
removed.